### PR TITLE
Enable light mode code in light mode

### PR DESCRIPTION
**Pull request by zhang1pr**
Pull request for issue #3253

_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making the changes.

**Summary**: 
1. Make class component `CodeBlock` render with prism's light theme in light mode.
2. Add light mode styles for short snippet expand line, long snippet expand button and gradient

I couldn't use context hook in the class so I passed the `isDarkMode` as a required prop from parent components. I really need help testing this change as I'm not sure where to test some components below on website and possibly missed something.

Besides the storybook file, there are 5 parent components using `CodeBlock` and affected by this PR. Test these components with Light Mode, Dark Mode and System:
1. `ProblemSolutions` - tested
2. `QuizGeneratorModal` - need help
3. `ProblemSubmissionPopup` - tested
4. `SafeMarkdownRenderer` - need help
5. `HTMLComponents` - tested

### CODE REVIEW

**Comment by zhang1pr** [comment]
I see that `CodeBlock.stories.tsx` isn't updated which is needed in the UI tests. Will update.

File: docs/Front End Documentation.md [hunk 1]
```diff
@@ -52,16 +54,16 @@ compile xdm into `src/gatsby/xdm.js` with Webpack:
...
-  entry: './index.js',
-  output: {
-    path: path.resolve('D:\\Code\\xdm', 'dist'),
-    filename: 'bundle.js',
-    library: {
-      name: 'xdm',
-      type: 'commonjs',
-    },
-  },
+	mode: 'production',
```
Comment (zhang1pr) [inline docs/Front End Documentation.md:57]: Done by linter.

File: package.json [hunk 2]
```diff
@@ -15,8 +15,8 @@
     "@heroicons/react": "^1.0.6",
     "@monaco-editor/react": "^4.4.5",
     "@sentry/gatsby": "^7.2.0",
-    "@storybook/builder-webpack5": "^6.5.9",
-    "@storybook/manager-webpack5": "^6.5.9",
+    "@storybook/builder-webpack5": "^6.5.16",
```
Comment (zhang1pr) [inline package.json:18]: Prompted by storybook for an upgrade. Any objection?

File: src/components/markdown/CodeBlock/CodeBlock.tsx [hunk 3]
```diff
@@ -44,9 +45,11 @@ const LineContent = styled.span`
 `;
 
 const CodeSnippetLineContent = styled(LineContent)`
-  color: #00bb00a0; // any better color?
```
Comment (zhang1pr) [inline src/components/markdown/CodeBlock/CodeBlock.tsx:47]: `#00bb00a0` isn't an accessible color that only has 4:1 contrast and WCAG AA asks for 4.5:1, so I removed the transparency.

File: src/components/markdown/CodeBlock/CodeBlock.tsx [hunk 4]
```diff
@@ -126,6 +129,7 @@ class CodeBlock extends React.Component<
   {
     children: string;
     className: string;
+    isDarkMode: boolean;
```
Comment (zhang1pr) [inline src/components/markdown/CodeBlock/CodeBlock.tsx:132]: I was debating required or optional, but finally opted for required.

File: src/components/markdown/CodeBlock/CodeBlock.tsx [hunk 5]
```diff
@@ -386,8 +411,9 @@ class CodeBlock extends React.Component<
                       style={
                         collapsed && isCodeBlockExpandable
                           ? {
-                              background:
-                                'linear-gradient(0deg, rgba(0,0,0,1) 0%, rgba(0,0,0,0) 100%)',
+                              background: isDarkMode
```
Comment (zhang1pr) [inline src/components/markdown/CodeBlock/CodeBlock.tsx:414]: Gradient styles.

File: src/components/markdown/CodeBlock/CodeBlock.tsx [hunk 6]
```diff
@@ -397,8 +423,9 @@ class CodeBlock extends React.Component<
                         viewBox="0 0 24 24"
                         stroke="currentColor"
                         className={
-                          'text-white w-6 h-6 transform group-hover:-translate-y-2 transition mb-2 ' +
-                          (collapsed ? '' : 'rotate-180')
+                          'w-6 h-6 transform group-hover:-translate-y-2 transition mb-2 ' +
```
Comment (zhang1pr) [inline src/components/markdown/CodeBlock/CodeBlock.tsx:426]: Expand button styles.

File: package.json [hunk 7]
```diff
@@ -15,8 +15,8 @@
     "@heroicons/react": "^1.0.6",
     "@monaco-editor/react": "^4.4.5",
     "@sentry/gatsby": "^7.2.0",
-    "@storybook/builder-webpack5": "^6.5.9",
-    "@storybook/manager-webpack5": "^6.5.9",
+    "@storybook/builder-webpack5": "^6.5.16",
+    "@storybook/manager-webpack5": "^6.5.16",
```
Comment (zhang1pr) [inline package.json:19]: Prompted by storybook for an upgrade. Any objection?

**Comment by SansPapyrus683** [comment]
wait so initially when the guide was in light mode, the code was still in dark mode?

**Comment by zhang1pr** [comment]
> wait so initially when the guide was in light mode, the code was still in dark mode?

Yes, code blocks had been always rendered in prism's `vs-dark` theme, now they are using `vs-light` theme in light mode.

File: package.json (hunk 7 above)
Comment (thecodingwizard) [reply package.json:19]: if you updated package.json, shouldn't `yarn.lock` change as well?

Comment (zhang1pr) [reply package.json:19]: Good point, `yarn.lock` did change and I thought I didn't need to check in that file. I just included it on my branch.

**Comment by bqi343** [comment]
personally, I prefer dark mode code when using light mode - I find it odd how there's no boundary between the code and the surrounding text when the code is light mode:

Screen Shot 2023-02-10 at 11 26 22 PM

Maybe add an option to toggle between light mode and dark mode code in settings?

**Comment by zhang1pr** [comment]
@bqi343 Good catch, I missed adding a background colour to the code block. It should look like this and I will change the `copy` tag colour as well.

Screenshot 2023-02-11 at 3 27 57 PM

Does that work for you instead of adding an option to change the code editor theme?

**Comment by bqi343** [comment]
That works

**Comment by thecodingwizard** [comment]
A lighter shade of gray might look better? Something like https://www.joshwcomeau.com/react/usememo-and-usecallback/#:~:text=This%20is%20precisely%20what%20useMemo%20allows%20us%20to%20do.%20Here%27s%20what%20it%20looks (what are your thoughts?)

**Comment by zhang1pr** [comment]
> A lighter shade of gray might look better? Something like https://www.joshwcomeau.com/react/usememo-and-usecallback/#:~:text=This%20is%20precisely%20what%20useMemo%20allows%20us%20to%20do.%20Here%27s%20what%20it%20looks (what are your thoughts?)

That indeed looks better. I just found I could use Tailwind's `bg-gray-50` so that colours are consistent and it fits collapsable code block like this (the code block background is white before and top background has been `bg-gray-50`)
Screenshot 2023-02-14 at 3 06 30 PM

**Review by thecodingwizard — APPROVED** [review]
completely untested oops

**Comment by zhang1pr** [comment]
Should have made it clear that I haven't updated the gray background colour yet. Will update.

**Comment by SansPapyrus683** [comment]
ayo what

**Comment by zhang1pr** [comment]
Oh my bad, I was trying to sync the branches. And GitHub detected no change on my branch for a second and auto-closed the pull request. Reopen the pull request now.

**Comment by zhang1pr** [comment]
Oops, try reopen again.

**Comment by zhang1pr** [comment]
This is ready for review again. Here are 2 recent changes:
1. Add a light mode gray background colour to code blocks as [discussed above](https://github.com/cpinitiative/usaco-guide/pull/3405#issuecomment-1430514253).
I had to override the rule from the prism theme rule in Highlight component.

2. Add light mode styles for copy buttons (static and hover) and language tags so they have enough contrast on the gray background. For the language tags, I tried to darken the original warm colours but they did not look nice and one of them looks very close to the copy button, so I tried to invert them and the cold colours look great. Hope this is not a problem.

Here are the new light mode look and the unchanged dark mode look.
Screenshot 2023-02-20 at 10 59 44 PM
Screenshot 2023-02-20 at 11 00 36 PM

**Review by bqi343 — APPROVED** [review]

**Pull request merged**